packages/ripple-keypairs/src/index.ts

```diff
@@ -38,66 +46,79 @@ const secp256k1 = {
-        .getPublic()
-        .encodeCompressed(),
-    )
+    const publicKey = bytesToHex(nobleSecp256k1.getPublicKey(derived, true))
     return { privateKey, publicKey }
   },
 
-  sign(message, privateKey): string {
-    return bytesToHex(
-      Secp256k1.sign(hash(message), hexToBytes(privateKey), {
-        canonical: true,
-      }).toDER(),
+  sign(message: ByteArray, privateKey: HexString): string {
+    // Some callers pass the privateKey with the prefix, others without.
+    // elliptic.js implementation ignored the prefix, interpreting it as a
+    // leading zero byte. @noble/curves will throw if the key is not exactly
+    // 32 bytes, so we normalize it before passing to the sign method.
+    // TODO: keep back compat like this, or simply always require prefix as
+    // the ed25519 sign method does.
+    assert.ok(
+      (privateKey.length === 66 && privateKey.startsWith(SECP256K1_PREFIX)) ||
+        privateKey.length === 64,
     )
+    const normed = privateKey.length === 66 ? privateKey.slice(2) : privateKey
+    return nobleSecp256k1
+      .sign(hash(message), normed)
+      .toDERHex(true)
+      .toUpperCase()
   },
 
   verify(message, signature, publicKey): boolean {
-    return Secp256k1.verify(hash(message), signature, hexToBytes(publicKey))
+    const decoded = nobleSecp256k1.Signature.fromDER(signature)
+    return nobleSecp256k1.verify(decoded, hash(message), publicKey)
   },
 }
 
 const ed25519 = {
-  deriveKeypair(entropy: Uint8Array): {
+  deriveKeypair(entropy: ByteArray): {
     privateKey: string
     publicKey: string
   } {
     const prefix = 'ED'
     const rawPrivateKey = hash(entropy)
     const privateKey = prefix + bytesToHex(rawPrivateKey)
     const publicKey =
-      prefix + bytesToHex(Ed25519.keyFromSecret(rawPrivateKey).pubBytes())
+      prefix + bytesToHex(nobleEd25519.getPublicKey(rawPrivateKey))
     return { privateKey, publicKey }
   },
 
-  sign(message, privateKey): string {
-    // caution: Ed25519.sign interprets all strings as hex, stripping
-    // any non-hex characters without warning
-    if (!Array.isArray(message)) {
-      throw new Error('message must be array of octets')
-    }
+  sign(message: ByteArray, privateKey: HexString): string {
+    assert.ok(
+      Array.isArray(message) || message instanceof Uint8Array,
+      'message must be array of octets',
+    )
+    assert.ok(
+      privateKey.length === 66,
+      'private key must be 33 bytes including prefix',
+    )
     return bytesToHex(
-      Ed25519.sign(message, hexToBytes(privateKey).slice(1)).toBytes(),
+      nobleEd25519.sign(new Uint8Array(message), privateKey.slice(2)),
     )
   },
 
-  verify(message, signature, publicKey): boolean {
-    return Ed25519.verify(
-      message,
-      hexToBytes(signature),
-      hexToBytes(publicKey).slice(1),
+  verify(
+    message: ByteArray,
+    signature: HexString | Uint8Array,
+    publicKey: string,
+  ): boolean {
```

Comment:
Should this check the length of the publicKey before slicing to give a cleaner error message if the public key has the wrong number of bytes?

Reply:
It won't get past getAlgorithmFromKey but it can't hurt to be defensive consistently

Reply:
Done in https://github.com/XRPLF/xrpl.js/pull/2273/commits/48266c39e46f0d1e5964463753e04adbb4f764e1

Reply:
I think it should fall through.  Turns out adding an exception here is unreachable with shorter public key.  The ed25519 is not exported and the only way to access its methods is through `verify` which calls `getAlgorithmFromKey` like sublimator mentioned.

This was the unit test I tried

```
    assert.throws(() => {
      api.verify(messageHex, signature, publicKey.slice(2))
    }, /^Error: public key must be 33 bytes including prefix'$/)
```

Instead it will fail with the error `Error: Invalid signature tag` because it is being treated as a secp256k1 key.

Reply:
Ooh, good point!

Maybe it should be split up into getPublicKeyAlgorithm/getPrivateKeyAlgorithm ?
The length checks could be put there ?

Reply:
I updated getAlgorithmFromKey to take a `type: 'public' | 'private'` parameter

Reply:
Why not have separate functions where that's a helper function?